aten/src/ATen/native/LossNLL2d.cpp

```diff
@@ -0,0 +1,356 @@
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/TensorUtils.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline void check_inputs_nll_loss2d(
+    const Tensor& input,
+    const Tensor& target,
+    const Tensor& weight) {
+  TORCH_CHECK(
+      target.dim() == 3,
+      "only batches of spatial targets supported (3D tensors)"
+      " but got targets of dimension: ",
+      target.dim());
+  TORCH_CHECK(
+      input.dim() == 4,
+      "only batches of spatial inputs supported (4D tensors), "
+      "but got input of dimension: ",
+      input.dim());
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == input.size(1),
+      "weight tensor should be defined either for all or no classes");
+
+  const int64_t input0 = input.size(0);
+  const int64_t input2 = input.size(2);
+  const int64_t input3 = input.size(3);
+  const int64_t target0 = target.size(0);
+  const int64_t target1 = target.size(1);
+  const int64_t target2 = target.size(2);
+  TORCH_CHECK(
+      input0 == target0 && input2 == target1 && input3 == target2,
+      "size mismatch (got input: ",
+      input.sizes(),
+      " , target: ",
+      target.sizes());
+}
+
+inline void check_gradout_shape_nll_loss2d(
+    const Tensor& grad_output,
+    const Tensor& target) {
+  TORCH_CHECK(
+      grad_output.dim() == 3,
+      "grad_output must have same dimension as target (3) but got dimension: ",
+      grad_output.sizes());
+
+  const int64_t grad_output0 = grad_output.size(0);
+  const int64_t grad_output1 = grad_output.size(1);
+  const int64_t grad_output2 = grad_output.size(2);
+  const int64_t target0 = target.size(0);
+  const int64_t target1 = target.size(1);
+  const int64_t target2 = target.size(2);
+  TORCH_CHECK(
+      grad_output0 == target0 && grad_output1 == target1 &&
+          grad_output2 == target2,
+      "size mismatch (got grad_output: ",
+      grad_output.sizes(),
+      " target: ",
+      target.sizes());
+}
+
+void nll_loss2d_forward_out_cpu_template(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction,
+    int64_t ignore_index) {
+  check_inputs_nll_loss2d(input, target, weight);
+  const int64_t n_classes = input.size(1);
+  total_weight.resize_({1});
+
+  AT_DISPATCH_FLOATING_TYPES_AND(
+      ScalarType::BFloat16,
+      input.scalar_type(),
+      "nll_loss2d_forward_out_cpu_template",
+      [&] {
+        auto weight_contiguous = weight.contiguous();
```

Comment:
why do we need to make it contiguous?

Reply:
I made it contiguous to access it directly via data-pointer. The problem is when using `TensorAccessor` I get an error when I call `access()` when `!weight.defined()` (I think it tries to get the size of dimensions of the tensor which is not possible when it is not defined). I tried that first but I found no really elegant solution that compared to the `weight_data == nullptr ? ...` ternary statements. 
Maybe I could use a pointer to an TensorAccessor but I thought that would be worse than just making weight contiguous and accessing the elements via `weighdata_ptr<scalar_t>()`.

Reply:
(calling `weight.contiguous();` is possible on undefined tensors)